lib/templates.php

```diff
@@ -155,3 +155,55 @@ function gutenberg_render_template_list_table_column( $column_name, $post_id ) {
 	echo esc_html( $post->post_name );
 }
 add_action( 'manage_wp_template_posts_custom_column', 'gutenberg_render_template_list_table_column', 10, 2 );
+
+/**
+ * Filter for adding a `resolved` parameter to `wp_template` queries.
+ *
+ * @param array $query_params The query parameters.
+ * @return array Filtered $query_params.
+ */
+function filter_rest_wp_template_collection_params( $query_params ) {
+	$query_params += array(
+		'resolved' => array(
+			'description' => __( 'Whether to filter for resolved templates', 'gutenberg' ),
+			'type'        => 'boolean',
+		),
+	);
+	return $query_params;
+}
+apply_filters( 'rest_wp_template_collection_params', 'filter_rest_wp_template_collection_params', 99, 1 );
+
+/**
+ * Filter for supporting the `resolved` parameter in `wp_template` queries.
+ *
+ * @param array           $args    The query arguments.
+ * @param WP_REST_Request $request The request object.
+ * @return array Filtered $args.
+ */
+function filter_rest_wp_template_query( $args, $request ) {
+	if ( $request['resolved'] ) {
+		$template_ids  = array( 0 ); // Return nothing by default (the 0 is needed for `post__in`).
+		$template_type = $request['slug'] ? array( $request['slug'] ) : get_template_types();
+
+		foreach ( get_template_types() as $template_type ) {
+			// Skip 'embed' for now because it is not a regular template type.
+			// Skip 'index' because it's a fallback that we handle differently.
+			if ( in_array( $template_type, array( 'embed', 'index' ), true ) ) {
+				continue;
+			}
+
+			// TODO: gutenberg_find_template_post_and_parts() internally performs a query.
+			// This means that there's room for optimization, and we might be able to
+			// generalize the logic found here into something that runs upon every `wp_template`
+			// query (not just those coming from the REST API).
```

Comment:
Why would we want that? We still need a way to use the regular querying logic.

Reply:
I was thinking of adding some kind of flag, similar to the `resolved` bool we're adding to the REST API endpoint in this PR. The same technique might not carry over 1:1 to `WP_Query`, but maybe we could add e.g. a dummy meta field.

I think that we _could_ consider manipulating _all_ queries for `wp_template`. If you look at the querying and auto-draft creation logic in `gutenberg_find_template_post_and_parts`, and where it's used, it's arguable that we'll always want to create those auto-drafts when querying a `wp_template`.

Reply:
But, you need to query the templates to figure out which ones you need to create auto-drafts for. The only way I see to get around that is to generate auto-drafts for all files before queries are made. Still, then we have to make sure we always update them when the underlying files change and that we give priority to actual customized non-auto-draft counterparts. All in all, I don't see it being much more straightforward, and we introduce a lot of magic into the very standardized functionality of `WP_Query`.

Reply:
> But, you need to query the templates to figure out which ones you need to create auto-drafts for. The only way I see to get around that is to generate auto-drafts for all files before queries are made. 

That depends on the query, doesn't it? If there's no `post_name` or `post_name__in` param, then that's true -- but that means it'd be pretty similar to what we have now in `edit-site-page.php` or in the new REST API filter I'm adding here -- where we just loop over template hierarchies.

> Still, then we have to make sure we always update them when the underlying files change and that we give priority to actual customized non-auto-draft counterparts.

That should be feasible by splitting what we now have in `gutenberg_find_template_post_and_parts` along sensible lines (like I sketched [here](https://github.com/WordPress/gutenberg/pull/21981#discussion_r423962066)), no?

> All in all, I don't see it being much more straightforward, and we introduce a lot of magic into the very standardized functionality of `WP_Query`.

Yeah, I think the magic vs. expected behavior is the biggest concern.

Reply:
Yeah, I don't see it as a definite net improvement.

We can explore it, but I wouldn't prioritize it.